### PR TITLE
[Clang][LLVM][Coroutines] Prevent __coro_gro from outliving __promise

### LINKED ISSUE #49843: Coroutine use after free on Clang 12.
|  |  |
| --- | --- |
| Bugzilla Link | [50499](https://llvm.org/bz50499) |
| Version | 12.0 |
| OS | All |
| CC | @adrianimboden,@zygoloid |

## Extended Description 
Starting with Clang 12, the address sanitizers fail because of a use-after-free any time `get_return_object` returns something that is not the parent object and the parent object is not empty. Here's a reproducer: https://godbolt.org/z/6Gfv3xG9o

Refs https://reviews.llvm.org/D103196 where this came up.

### PR DESCRIPTION
**Pull request by bcardosolopes**
When dealing with short-circuiting coroutines (e.g. expected), the deferred calls that resolve the get_return_object are currently being emitted after we delete the coroutine frame.

This was caught by ASAN when using optimizations -O1 and above: optimizations after inlining would place the __coro_gro in the heap, and subsequent delete of  the coroframe followed by the conversion -> BOOM.

This patch forbids the GRO to be placed in the coroutine frame, by adding a new metadata node that can be attached to `alloca` instructions.

Fix #49843.

### CODE REVIEW

**Comment by ChuanqiXu9** [comment]
I remember that there is a defect that we may place the GRO on the coroutine frame. And my instinct reaction is that would this patch be covered by forcing GRO to not live on the coroutine frame?

**Comment by bcardosolopes** [comment]
Thanks for the fast reply @ChuanqiXu9 

> I remember that there is a defect that we may place the GRO on the coroutine frame.

Can you point me to such defect? I had no luck searching for it.

> And my instinct reaction is that would this patch be covered by forcing GRO to not live on the coroutine frame?

How do you suggest we do so? Even if we teach the optimizers to not touch the GRO (e.g. `-O0`) there's still a correctness issue. The generated code is still wrong, given that currently, the lifetime of `__coro_gro` outlives the `__promise`, but accessing the proxy for delayed conversion might still depend on the promise.

Also, what do you mean by "patch be covered"? I'm confused.

**Comment by bcardosolopes** [comment]
Updated the patch to account for failing libcxx tests, and to not change current codegen when GRO allocas are not involved.

**Comment by ChuanqiXu9** [comment]
> Can you point me to such defect? I had no luck searching for it.

It should be https://github.com/llvm/llvm-project/issues/49843. I didn't expect that it was not marked as coroutines. Besides fixing that bug, it is also a minor optimization to not include GRO in coroutine frames.

> Also, what do you mean by "patch be covered"? I'm confused.

When I say "patch be covered", I mean the problem may be solved automatically if we can force the GRO not living on the coroutine frames.

> How do you suggest we do so? Even if we teach the optimizers to not touch the GRO (e.g. -O0) there's still a correctness issue. The generated code is still wrong, given that currently, the lifetime of __coro_gro outlives the __promise, but accessing the proxy for delayed conversion might still depend on the promise.

By forcing the GRO not living on the coroutine frames, it shouldn't be a problem if the lifetime of `__coro_gro` outlives the `__promise`. The only limit is that the initialization of `__coro_gro` should be in the lifetime of `__promise`.

**Comment by bcardosolopes** [comment]
Thanks for the clarifications

> By forcing the GRO not living on the coroutine frames, it shouldn't be a problem if the lifetime of `__coro_gro` outlives the `__promise`. The only limit is that the initialization of `__coro_gro` should be in the lifetime of `__promise`.

Applied your suggestion based on #49843, and edited description + title.

File: clang/test/CodeGenCoroutines/coro-gro.cpp [hunk 1]
```diff
@@ -104,3 +105,5 @@ invoker g() {
   // CHECK: call void @_ZN7invoker15invoker_promise17get_return_objectEv({{.*}} %[[AggRes]]
   co_return;
 }
+
+// CHECK: ![[OutFrameMetadata]] = !{}
```
Comment (ChuanqiXu9) [inline clang/test/CodeGenCoroutines/coro-gro.cpp:109]: nit: no newline

File: llvm/docs/LangRef.rst [hunk 2]
```diff
@@ -6691,6 +6691,22 @@ sections that the user does not want removed after linking.
...
+Additionally, this metadata is only used as a flag, so the associated
+node must be empty.
+
+.. code-block:: text
+
+  %__coro_gro = alloca %struct.GroType, align 1, !coro.outside.frame !0
+
+  ...
+  !0 = !{}
+
```
Comment (ChuanqiXu9) [inline llvm/docs/LangRef.rst:6709]: Maybe it is better to add this to a new section of  `Coroutines.rst`. We add the coroutines related things there generally.

**Review by ChuanqiXu9 — COMMENTED** [review]
Thanks. It looks good if we can add a llvm test in llvm/test/Transforms/Coroutines for the new metadata.

File: llvm/docs/LangRef.rst (hunk 2 above)
Comment (bcardosolopes) [reply llvm/docs/LangRef.rst]: Done!

File: clang/test/CodeGenCoroutines/coro-gro.cpp (hunk 1 above)
Comment (bcardosolopes) [reply clang/test/CodeGenCoroutines/coro-gro.cpp:108]: Done

**Comment by bcardosolopes** [comment]
Applied suggested changes and updated PR.

**Comment by apolloww** [comment]
Thanks for the fix! I think originally the deferred conversion is placed after `coro.end` so that it won't be counted as "use across suspend point", but it doesn't stop optimizer from breaking that assumption.

**Review by ChuanqiXu9 — APPROVED** [review]
LGTM. Thanks.

**Pull request merged**